File: tictactoe.py
import random
import time 

def set_chess_location(x, y, chess_board, user):
    if user == 0:
        chess_mark = 'O'
    elif user == 1:
        chess_mark = 'X'
    if chess_board[x][y] == 'X' or chess_board[x][y] == 'O':
        return 0
    chess_board[x][y] = chess_mark
    return 1


def get_chess_location(rawDate):
    temp = rawDate
    if temp <= 9:  # 顺数坐标
        if temp/3 <= 1:  # 第一行
            m = 0
            n = temp % 3
            n -= 1
        elif 1 < temp/3 <= 2:
            m = 1
            n = temp % 3
            n -= 1
        elif 2 < temp/3 <= 3:
            m = 2
            n = temp % 3
            n -= 1
    else:
        m, n = tuple(str(temp))
        m = int(m)-1  # 为数组做的处理
        n = int(n)-1
    return (m, n)


def whoIsWiner(chess):
    if chess == 'O':
        print("玩家1胜利")
    elif chess == 'X':
        print('玩家2胜利')


def iSContinueGame(flag, chess_board):
    if flag == 0:
        print('游戏结束')
        input('任意键继续')
        return 1
    return 0


def chess_robot(diffcult_flag,chess_board):
    user_flag = 1
    if diffcult_flag == 1:
       while True:
            randomNum = lambda:random.randint(0,2)
            m,n = (randomNum(),randomNum())
            result = set_chess_location(m,n,chess_board,user_flag)
            if result == 1:
                time.sleep(1)
                print('机器人下在了{}，{}'.format(m+1,n+1))
                break
    else:
        pass

    return result

def chess_user(chess_board,user_flag,temp):
    temp = int(temp)
    m,n = get_chess_location(temp)
    state = set_chess_location(m,n,chess_board,user_flag)#返回棋子安放信息，是否下在了同一位置
    return state

def judgeResult(chess_board):
    flag = 0  # 代表游戏是否还可以继续  0 代表不可以（赢了或者和棋了）1代表还未和棋或赢
    for m in range(3):
        if chess_board[m][0] == chess_board[m][1] == chess_board[m][2] and chess_board[m][2] != 0:
            whoIsWiner(chess_board[m][0])
            flag = 0
            break
        elif chess_board[0][m] == chess_board[1][m] == chess_board[2][m] and chess_board[2][m] != 0:
            whoIsWiner(chess_board[m][0])
            flag = 0
            break
        else:
            flag = 1
    if flag == 1:  # 在上一种赢法下游戏还没有结束可以继续 反之游戏结束
        if chess_board[0][0] == chess_board[1][1] == chess_board[2][2] and chess_board[1][1] != 0:
            whoIsWiner(chess_board[1][1])
            flag = 0
        elif chess_board[0][2] == chess_board[1][1] == chess_board[2][0] and chess_board[1][1] != 0:
            whoIsWiner(chess_board[1][1])
            flag = 0
    flag_init_chess_board = iSContinueGame(flag, chess_board)
    if flag == 1:  # 在得不到输赢的情况下检查是否已经和棋
        flag = 0 #变为默认离开游戏，由chess的值来阻止
        for row in chess_board:
            for chess in row:
                if chess == 0:
                    flag = 1
        flag_init_chess_board = iSContinueGame(flag, chess_board)
    return flag_init_chess_board


def isInputLegal(num):
    try:
        num = int(num)
    except:
        return 0
    if 0 < num <= 13:
        return 1
    elif 21 <= num <= 23:
        return 1
    elif 31 <= num <= 33:
        return 1
    return 0


def chess_board_print(chess_board):
    for row in chess_board:
        for chess in row:
            if chess != 0:
                print(chess, end=' ')
            else:
                print("？", end=" ")
        print('\n')


def main():
    chess_board = [[0]*3 for i in range(3)]
    while True:
        robot_flag = input('输入1选择人机模式，输入2选择玩家对战')
        if robot_flag == '1' or robot_flag =='2':
            robot_flag = int(robot_flag)
            if robot_flag == 2:
                break
            diffcult_flag = input('输入1选择简单模式，输入2选择困难')
            if diffcult_flag == '1' or diffcult_flag == '2':
                diffcult_flag = int(diffcult_flag)
                break
        print('检查你的输入')
    chess_board_print(chess_board)
    user_flage = 0
    print("输入你要落字的位置可以是顺数位置也可以是坐标位置\n，行\
            ，列\n例如第二行第三列表示为23")

    while(True):
        if robot_flag == 1 and user_flage == 0:
            temp = input('玩家{},使用{}你选择的位置是：'.format(user_flage+1,'O' if user_flage == 0 else 'X'))
            if isInputLegal(temp) == 0:
                print("请检查输入是否合法", end='',flush=True)
                continue
            state = chess_user(chess_board,user_flage,temp)
        elif robot_flag == 1 and user_flage == 1:
            state = chess_robot(diffcult_flag,chess_board)
        elif robot_flag == 2:
            temp = input('玩家{},使用{}你选择的位置是：'.format(user_flage+1,'O' if user_flage == 0 else 'X'))
            if isInputLegal(temp) == 0:
                print("请检查输入是否合法", end='',flush=True)
                continue
            state = chess_user(chess_board,user_flage,temp)


        if user_flage == 1 and state == 1:
            user_flage = 0
            chess_board_print(chess_board)

        elif user_flage == 0 and state == 1:
            user_flage = 1
            chess_board_print(chess_board)
        else:
            print('输入位置已存在棋子，重新落子', end='',flush=True)
        if judgeResult(chess_board) == 1:
            chess_board = [[0]*3 for i in range(3)]


if __name__ == "__main__":
    main()

File: tictactoe.cpp
#include <stdio.h>
#include <time.h>
#include <stdlib.h>

int *translateLocation(int rawLocation)
{
    int m, n;
    float temp;
    static int x_y[2];
    temp = (float)rawLocation/3;
    /*这里所作的减1操作均是把习惯上1为下标表达转化成数组上0为下标的表达*/
    if (rawLocation <= 9)
    {
        if (temp <= 1)
        {
            m = 0;
        }
        else if ((1 < temp) && (temp <= 2))
        {
            m = 1;
        }
        else if ((2 < temp) && (temp <= 3))
        {
            m = 2;
        };
        n = rawLocation % 3;
        if (n == 0){
            n = 2;
        }
        else
        {
            n -= 1;
        }

    }
    else
    {
        n = rawLocation % 10;
        m = (rawLocation - n) / 10;
        n -= 1;
        m -= 1;
    };
    x_y[0] = m;
    x_y[1] = n;
    return x_y;
};
int set_chess_localtion(int rawLocation, int user_flage, int chess_board[][3])
{
    int x, y, *x_y;
    int chess_mark;
    x_y = translateLocation(rawLocation);
    x = *(x_y + 0);
    y = *(x_y + 1);
    if (user_flage == 0)
    {
        chess_mark = 'O';
    }
    else if (user_flage == 1)
    {
        chess_mark = 'X';
    };
    if (chess_board[x][y] == 'X' || chess_board[x][y] == 'O')
    {
        return 0;
    }
    printf("玩家%d已落子在行%d列%d\n",user_flage+1,x+1,y+1);
    chess_board[x][y] = chess_mark;

    return 1;
};
void whoIsWiner(int chess)
{
    if (chess == 'O')
    {
        printf("玩家1胜利\n");
    }
    else if (chess == 'X')
    {
        printf("玩家2胜利\n");
    }
};

int judgeResult(int chess_board[][3])
{
    int go_ahead_flag, isGameOver;
    for (int i = 0; i < 3; i++)
    {
        if ((chess_board[i][0] == chess_board[i][1]) && (chess_board[i][1] == chess_board[i][2]) && chess_board[i][1] != '?')
        {
            whoIsWiner(chess_board[i][0]);
            go_ahead_flag = 0;
            break;
        }
        else if ((chess_board[0][i] == chess_board[1][i]) && (chess_board[1][i] == chess_board[2][i]) && chess_board[1][i] != '?')
        {
            whoIsWiner(chess_board[0][i]);
            go_ahead_flag = 0;
            break;
        }
        else
        {
            go_ahead_flag = 1;
        }
    }
    if (go_ahead_flag == 1)
    {
        if ((chess_board[0][0] == chess_board[1][1]) && (chess_board[1][1] == chess_board[2][2]) && chess_board[1][1] != '?')
        {
            whoIsWiner(chess_board[1][1]);
            go_ahead_flag = 0;
        }
        else if ((chess_board[0][2] == chess_board[1][1]) && (chess_board[1][1] == chess_board[2][0]) && chess_board[1][1] != '?')
        {
            whoIsWiner(chess_board[1][1]);
            go_ahead_flag = 0;
        }
    }
    if (go_ahead_flag == 1)
    {
        go_ahead_flag = 0;
        for (int i = 0; i < 3; i++)
        {
            for (int j = 0; j < 3; j++)
            {
                if (chess_board[i][j] == '?')
                {
                    go_ahead_flag = 1;
                    break;
                }
            }
            if (go_ahead_flag == 1)
            {
                break;
            }
        }
    };
    isGameOver = !(go_ahead_flag);
    return isGameOver;
};

int isInputLegal(int input)
{
    if (0 < input && input <= 13)
    {
        return 1;
    }
    else if (21 <= input && input <= 23)
    {
        return 1;
    }
    else if (31 <= input && input <= 33)
    {
        return 1;
    }
    return 0;
}

void chess_board_print(int chess_board[3][3]) /*让棋盘显示出来，好看*/
{
    printf("%c|%c|%c\n%c|%c|%c\n%c|%c|%c\n",
           chess_board[0][0],
           chess_board[0][1],
           chess_board[0][2],
           chess_board[1][0],
           chess_board[1][1],
           chess_board[1][2],
           chess_board[2][0],
           chess_board[2][1],
           chess_board[2][2]);
}

int *get_how_to_game()
{
    int robot_flag, isRobotFlagLegal, isDiffcultFlagLegal, diffcult_flag;
    static int temp[2];
    while (1)
    {
        printf("输入1选择玩家对战，输入2选择人机对战\n");
        isRobotFlagLegal = scanf("%d", &robot_flag);
        if (isRobotFlagLegal == 0)
        {
            continue;
        }
        else
        {
            robot_flag -= 1; /*处理一下使得1为人机对战，0为玩家对战*/
            if (robot_flag == 1)
            {
                printf("输入1选择简单模式，2为困难模式");
                isDiffcultFlagLegal = scanf("%d", &diffcult_flag);
                if (isDiffcultFlagLegal != 0)
                {
                    if (diffcult_flag == 1 || diffcult_flag == 2)
                    {
                        break;
                    }
                }
            }
            else if (robot_flag == 0)
            {
                break;
            }
        }
    };
    temp[0] = robot_flag;
    temp[1] = diffcult_flag;
    return temp;
}

int get_where_to_go(int robot_flag, int user_flage)
{
    int where_to_go, isOneOr2;
    int x, y,temp1,temp2;
    while (1)
    {
        if ((robot_flag == 1 && user_flage == 0) || (robot_flag == 0))
        {
            printf("玩家%d,使用%c，你选择的位置是:\n",
                   user_flage + 1,
                   user_flage == 0 ? 'O' : 'X');
            isOneOr2 = scanf("%d", &where_to_go);
            if (isInputLegal(where_to_go) == 1)
            {
                break;
            }
            printf("请检查输入是否合法\n");
        }
        else
        {
            /*Easy机器人下棋*/
            srand((unsigned)time(NULL));
            temp1 = rand() % 3;
            temp2 = rand() % 3;
            temp1 == 0?x = 3:(x = temp1);
            temp2 == 0?y = 3:(y = temp2);
            where_to_go = (x * 10) + y; /*对数据进行统一处理*/
            break;
        }
    };
    return where_to_go;
}

void game(int *game_model, int chess_board[][3])
{
    int user_flage = 0;
    int robot_flag, diffcult_flag, isOneOr2, set_chess_state;
    int where_to_go;
    int black_hole;
    robot_flag = *(game_model + 0);
    diffcult_flag = *(game_model + 1);
    while (1)
    {
        where_to_go = get_where_to_go(robot_flag, user_flage);
        set_chess_state = set_chess_localtion(where_to_go, user_flage, chess_board);
        if (set_chess_state == 1)
        {
            user_flage = !(user_flage);
            chess_board_print(chess_board);
        }
        else
        {   
            if (robot_flag == 1 && user_flage == 1){
                continue;
            }
            printf("该位置已经存在落子，请重新选取位置");
        };
        if (judgeResult(chess_board) == 1)
        {
            printf("游戏结束\n任意字符回车\n开始下一局");
            scanf("%d",&black_hole);
            for (int i = 0; i < 3; i++)
            {
                for (int j = 0; j < 3; j++)
                {
                    chess_board[i][j] = '?';
                }
            };
            chess_board_print(chess_board);
        }
    }
};
int main()
{
    int chess_board[3][3] = {
        '?',
        '?',
        '?',
        '?',
        '?',
        '?',
        '?',
        '?',
        '?',
    };
    int *game_model;
    game_model = get_how_to_game();
    /*    printf("%d,%d",*(game_model+0),*(game_model+1));  */
    chess_board_print(chess_board);
    printf("输入你要落字的位置可以是顺数位置也可以是坐标位置\n，行\
            ，列\n例如第二行第三列表示为23\n");
    game(game_model, chess_board);
}
